### ISSUE TITLE
Quickbooks source: Refresh token expires after 1 hour in Cloud

### ISSUE THREAD

**Issue by natikgadzhi** [airbytehq/airbyte#44471]
## Connector Name
source-quickbooks

## ❓ What
The Quickbooks connector fails after the refresh token expires, as the refresh token expires after 1 hour. 

## 📈  Impact
Quickbook syncs will be able to continue running without failures. Users often see the second sync and every sync after fail.

## 🤝 Proposed solution
Rework the authentication for the Quickbook OAuth so that it refreshes the expiry of the token. A refresh token can last up to 100 days.

**Comment by natikgadzhi** [comment]
This one is pretty high leverage. Let's make sure to implement token refresh mechanism. You will need test accounts though, we'd need to figure out a partner to give us their creds.

**Comment by csphereholdings** [comment]
Hey there, @natikgadzhi - happy to share some test creds in QB with you. Nothing sensitive there.

**Comment by pabloescoder** [comment]
Hey @csphereholdings, I'll be working on the issues you posted for QuickBooks - this one and the issue that you're getting about the eval method.

Could you please share the test QB credentials with me? 
Or alternatively if you prefer you can share it with @natikgadzhi and I can get those from him, anything works.

**Comment by csphereholdings** [comment]
Hello @pabloescoder aka Dhroov ;)

Just e-mailed you.

Cheers!

**Comment by csphereholdings** [comment]
Hello! @pabloescoder just wanted to check in and see how things are coming along? Anything I can help with?

**Comment by pabloescoder** [comment]
Hey hey @csphereholdings 
[I've created a PR](https://github.com/airbytehq/airbyte/pull/44560) for the QuickBooks error that you were getting while setting up the connector (str has no method eval) I've resolved that, and while I was testing that I noticed that the token was being refreshed automatically (I was able to get results with an access token that I had created the previous day - so Airbyte was using the refresh token to generate a new access token)
Anyway, we've kicked off regression tests for that connector, and I'm waiting for that PR to get merged, once that's done we can see if the issue with the token is resolved, if not then I'll look into why that's the case. 
CC: @natikgadzhi

**Comment by csphereholdings** [comment]
Awesome! Thanks so much!

If I can do anything, please do reach out!

Christian M. Grupp
CEO
Mobile: (307) 203.3008 | Email: ***@***.***
Sphere Holdings, 2810 N Church St PMB 36678, Wilmington, DE 19802

www.<http://www.my-company.com/>sphereholdings.co

[signatureImage]







________________________________

**Comment by pabloescoder** [comment]
Sure, will do, thank you!

**Comment by charles-astrafy** [comment]
What's the status on this issue ? The connector is a bit useless if the refresh token mechanism is not working.

Thank you.

**Comment by csphereholdings** [comment]
Greetings! Any updates here? Anything we can do to help close this out?

**Comment by pabloescoder** [comment]
@charles-astrafy @csphereholdings We've released a new version (3.0.26) that fixes a bug that made the connector unusable.
Are you seeing this issue again in 3.0.26?

**Comment by nawfel-bacha** [comment]
@pabloescoder I confirm that now it's working fine in 3.0.26, the token seems to be refreshed correctly.

**Comment by pabloescoder** [comment]
@charles-astrafy Awesome! I'm glad to hear that.
Closing issue as resolved in 3.0.26.

**Comment by nataliekwong** [comment]
Hi @pabloescoder , I noticed a new similar issue about the authentication expiring after 1 hour: https://github.com/airbytehq/airbyte/issues/49842. Would this issue be similar to the one you resolved or different?

Thanks,
Natalie

**Comment by pabloescoder** [comment]
Hey @nataliekwong, This issue should've been resolved in 3.0.26, but I can see that in the new issue that's raised, they're already on 3.0.26 and still seeing the issue.
 
The refresh token mechanism was working for me locally and also for Charles as mentioned above.
 
I'll investigate further to see why we're seeing this issue again.

**Comment by nataliekwong** [comment]
Got it, thanks!

**Comment by tomsweeting** [comment]
Hey @pabloescoder - I have encountering this same issue using Arbyte 1.3.0 + Quickbooks connector 3.0.26

The initial sync works fine, however after 1 hour once the initial auth token has expired, the sync will fail with error: `Failure in source: Unauthorized. Please ensure you are authenticated correctly.`

Log from Airbyte:
```
2025-01-01 12:56:58 source ERROR Encountered an exception while reading stream customers
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/abstract_source.py", line 133, in read
    yield from self._read_stream(
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/abstract_source.py", line 239, in _read_stream
    for record_data_or_message in record_iterator:
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/streams/core.py", line 192, in read
    for record_data_or_message in records:
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/declarative_stream.py", line 136, in read_records
    yield from self.retriever.read_records(self.get_json_schema(), stream_slice)
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 375, in read_records
    for stream_data in self._read_pages(record_generator, self.state, _slice):
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 298, in _read_pages
    response = self._fetch_next_page(stream_state, stream_slice, next_page_token)
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 273, in _fetch_next_page
    return self.requester.send_request(
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/requesters/http_requester.py", line 305, in send_request
    request, response = self._http_client.send_request(
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/streams/http/http_client.py", line 382, in send_request
    response: requests.Response = self._send_with_retry(
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/streams/http/http_client.py", line 228, in _send_with_retry
    response = backoff_handler(rate_limit_backoff_handler(user_backoff_handler))(request, request_kwargs, log_formatter=log_formatter, exit_on_rate_limit=exit_on_rate_limit)  # type: ignore # mypy can't infer that backoff_handler wraps _send
  File "/usr/local/lib/python3.10/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/airbyte_cdk/sources/streams/http/http_client.py", line 300, in _send
    raise MessageRepresentationAirbyteTracedErrors(
airbyte_cdk.sources.streams.http.http_client.MessageRepresentationAirbyteTracedErrors: Unauthorized. Please ensure you are authenticated correctly.
```

Happy to share additional information if it helps!

**Comment by tomsweeting** [comment]
I tested this again this morning to check if it was just a fluke and I can reproduce the problem:
- Generated new tokens ~ 10:30AM
- Ran three manual syncs within the hour, all succeeded
- Set up cron schedule
- On first schedule trigger at 12:01PM the connection no longer works

<img width="1231" alt="Image" src="https://github.com/user-attachments/assets/27f62d6e-0c3a-4272-8f40-f30b838d0c50" />

Can this issue be reopened, or should I post a new one?

**Comment by Fahmed225** [comment]
Hi @pabloescoder 

Did you get a chance to look into this? We're experiencing this on our end as well.

**Comment by pabloescoder** [comment]
Hey @Fahmed225 
We just released a new version of QB (v4.0.0) with the fix. 
It's a major version upgrade because the credentials object was removed from the config and all the config fields were moved to the root level so you might need to repopulate the config fields. 
TLDR: QB v4.0.0 solves this issue!
